Sources/AssociatedObjectPlugin/AssociatedObjectMacro.swift

```diff
@@ -328,14 +330,26 @@ extension AssociatedObjectMacro {
                     "let oldValue = \(identifier)"
                 }
 
-                """
-                setAssociatedObject(
-                    self,
-                    \(associatedKey),
-                    newValue,
-                    \(policy)
-                )
-                """
+                if type.isOptional, hasDefaultValue {
+                    """
+                    setAssociatedObject(
+                        self,
+                        \(associatedKey),
+                        newValue,
+                        \(policy)
+                    )
+                        self.__associated_\(identifier.trimmed)IsSet = true
+                    """
+                } else {
+                    """
+                    setAssociatedObject(
+                        self,
+                        \(associatedKey),
+                        newValue,
+                        \(policy)
+                    )
+                    """
+                }
```

Comment:
The following code may be better and less duplicative.

```suggestion
                """
                setAssociatedObject(
                    self,
                    \(associatedKey),
                    newValue,
                    \(policy)
                )
                """
                if type.isOptional, hasDefaultValue {
                    """
                    self.__associated_\(identifier.trimmed)IsSet = true
                    """
                }
```